crates/bevy_text/src/glyph_brush.rs

```diff
@@ -0,0 +1,178 @@
+                    let handle = &self.handles[section_glyph.font_id.0];
+                    let font = fonts.get(handle).ok_or(TextError::NoSuchFont)?;
+                    let glyph = &section_glyph.glyph;
+                    let scaled_font = ab_glyph::Font::as_scaled(&font.font, glyph.scale.y);
+                    // glyph.position.y = baseline
+                    max_y = max_y.max(glyph.position.y - scaled_font.descent());
+                    min_x = min_x.min(glyph.position.x - scaled_font.h_side_bearing(glyph.id));
+                }
+                max_y = max_y.floor();
+                min_x = min_x.floor();
+
+                for section_glyph in section_glyphs {
+                    let handle = &self.handles[section_glyph.font_id.0];
+                    let font = fonts.get(handle).ok_or(TextError::NoSuchFont)?;
+                    let glyph_id = section_glyph.glyph.id;
+                    let font_size = section_glyph.glyph.scale.y;
+                    if let Some(outlined_glyph) = font.font.outline_glyph(section_glyph.glyph) {
+                        let bounds = outlined_glyph.px_bounds();
+                        let handle_font_atlas: Handle<FontAtlasSet> = handle.as_weak();
+                        let font_atlas_set = font_atlas_set_storage
+                            .get_or_insert_with(handle_font_atlas, FontAtlasSet::default);
+
+                        let atlas_info = font_atlas_set
+                            .get_glyph_atlas_info(font_size, glyph_id)
+                            .map(Ok)
+                            .unwrap_or_else(|| {
+                                font_atlas_set.add_glyph_to_atlas(
+                                    texture_atlases,
+                                    textures,
+                                    outlined_glyph,
+                                )
+                            })?;
+
+                        let texture_atlas = texture_atlases.get(&atlas_info.texture_atlas).unwrap();
+                        let glyph_rect = texture_atlas.textures[atlas_info.glyph_index as usize];
+                        let glyph_width = glyph_rect.width();
+                        let glyph_height = glyph_rect.height();
+
+                        let x = bounds.min.x + glyph_width / 2.0 - min_x;
+                        // the 0.5 accounts for odd-numbered heights (bump up by 1 pixel)
+                        // max_y = text block height, and up is negative (whereas for transform, up is positive)
+                        let y = max_y - bounds.max.y + glyph_height / 2.0 + 0.5;
+                        let position = Vec2::new(x, y);
+
+                        vertices.push(TextVertex {
+                            position,
+                            atlas_info,
+                        });
+                    }
+                }
+                Ok(vertices)
+            })
+            .collect::<Result<Vec<_>, TextError>>()?
```

Comment:
Not necessarily something we need to do in this pr, but it would be interesting to try to remove allocations here entirely and just walk through each glyph + render it directly

Reply:
I do not think that this is feasible with the new design: we essentially store the TextGlyphs inside an entity now, and iterate over them. Collection needs to happen at some point, I believe.